### DISCUSSION TITLE
[Chapter 06] 퀴즈와 궁금한 점

### DISCUSSION THREAD

**Discussion by snoopingh** [cowaine/is-my-code-that-strange#54] [Chapter 06 (조건 분기)] [answered]
퀴즈) 부모 클래스의 메소드가 예외를 던지지 않도록 설계되었는데, 자식 클래스에서 해당 메소드를 오버라이딩 할 때, 예외를 던진다면 리스코프치환정책을 위반하는 것일까요?

질문) switch 로 분기처리 하는 코드를 인터페이스를 통한 전략패턴으로 구현하는게 객체지향적인 코딩이라는것에 감명 받았습니다.
p106에 인터페이스의 이름을 결정하는 방법과 관련된 내용이 있는데, 실무에서 이렇게 인터페이스로 뽑아 낸다는 것이 상당히 쉽지 않은 것 같습니다.

스터디원 분들은 이런 코딩 방법에 이미 익숙하신 걸까요?
인터페이스를 통한 구현으로 서비스 운영 단계에서 탁월한 효과를 본 사례가 있는지 궁금합니다.
그리고 많이 생성된 인터페이스나 클래스를 관리할 때, 패키지나 클래스 네임은 어떤식으로 관리하시는지도 궁금합니다.

**Comment by firewoody237** [comment]
오 좋은 퀴즈 같아요. 요건 정답이 있는 퀴즈일까요??
제 생각엔 리스코프 치환원칙은 "상위 타입을 하위 타입으로 교체해도 정상적으로 동작한다."의 정의 하에, 다음과 같은 이유로 위반한다고 생각합니다.
- (전제) A기능에 대해 부모 클래스에서는 예외를 던지지 않지만, 자식 클래스에선 예외를 던진다.
- 사용자의 "정상적" 판단을 벗어난다.
- 동일한 명칭의 메서드에 대한 처리 방법이 달라진다. 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
스프링의 컨테이너로 전략패턴은 이미 다들 사용하고 계시지 않을까요?
저도 추후 확장 및 변경이 확실한 경우 인터페이스를 두고 주입만 가끔 바꾼적이 있습니다.
클래스나 패키지 명은 잘 모르겠으나.. 가급적 `Impl` 접미사는 붙이지 않으려고 고뇌하는 것 같습니다 ㅎㅎ

**Comment by 0sunset0** [comment, marked as answer]
저도 Impl은 사용하지 않는 것 같아요 
역할 그대로를 이름에 담으려고 하는 것 같고.. 인터페이스랑 클래스는 같은 패키지에 둡니다!

비슷한 도메인이나 계층적인 도메인에서 인터페이스나 추상클래스를 사용할 수 있지 않을까하는 생각이 들었는데
김영한님이 [답변](https://www.inflearn.com/questions/63567/%EC%95%88%EB%85%95%ED%95%98%EC%8B%AD%EB%8B%88%EA%B9%8C-%EB%8F%84%EB%A9%94%EC%9D%B8-%EC%84%A4%EA%B3%84%EC%8B%9C-%EC%9D%B8%ED%84%B0%ED%8E%98%EC%9D%B4%EC%8A%A4-%EC%B6%94%EC%83%81%ED%81%B4%EB%9E%98%EC%8A%A4-%EC%97%90-%EB%8C%80%ED%95%9C-%EC%9D%98%EA%B2%AC%EC%9D%84-%EB%93%A3%EA%B3%A0-%EC%8B%B6%EC%8A%B5%EB%8B%88%EB%8B%A4)해주신거 보고 상황에 따라 결정해야 한다는 생각이 드네요 :)

글에서 할인 정책 얘기가 나오는데 사진 하나 첨부합니다!!
![image](https://github.com/cowaine/is-my-code-that-strange/assets/81086966/38f3899b-82ae-460c-a1a3-a5bb7a378a2b)

**Reply by snoopingh** [reply]
명확한 답이 되었습니다. 도메인과 밀접한 부분 중 인터페이스 설계가 여러운 부분에 대해서는 도대체 어떻게 해야 될 지 고민했었는데 안해도 되는 고민이었군요
